### PR TITLE
BI-1555 - Cache program ontology terms

### PR DESCRIPTION
**Pull request by timparsons**
# Description
**Story:** https://breedinginsight.atlassian.net/browse/BI-1555

- Modified the underlying cache storage from in memory of the running application to now being stored in Redis.  
- Added the Redisson client library
- Modified the `ProgramCache` to utilize Redis instead of Guava's `LoadingCache`.
- Implemented caching for the `TraitDAO`

One set of changes that appears rather irrelevant to the story is with `ProgramDAO` and `TraitDAO` being refactored to interfaces, and then the creation of `ProgramDAOImpl`, `TraitDAOImpl`, and `AbstractDAO`.  The main reason for this was when trying to mock the original `ProgramDAO` or `TraitDAO`, errors were thrown when creating the mocks.  The fix for this (per Micronaut) is to have interfaces for things you'd like to mock, especially if the class being mocked extends another class.  By creating an interface, mocking is much simpler, and doesn't run into any issues due to inheritance.

# Dependencies
Redis container -> this has been added to the `docker-compose.yml` file in bi-api, so you should be able to start up that container, and the code will default to `localhost:6379` to look for a Redis server to connect to.

# Testing
Test traits:
1. Start up bi-api and wait for a log message like `cache loading complete for key: <programUUID>:trait` (you may want to wait until you see your program UUID of interest)
2. navigate to the ontology terms page, and ensure the traits load (hopefully faster)

Test germplasm:
1. Start up bi-api and wait for a log message like `cache loading complete for key: <programUUID>:brapigermplasm` (you may want to wait until you see your program UUID of interest)
2. navigate to the all germplasm page, and ensure the germplasm records load (hopefully faster)

Test restart (records should persist in redis between bi-api restarts):
1. start bi-api and let the cache load
2. restart bi-api, and refresh the germplasm or trait list page immediatly after the server successfully starts up
3. ensure the page loads with all the appropriate records

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have tested my code and ensured it meets the acceptance criteria of the story
- [x] I have tested that my code works with both the brapi-java-server and BreedBase
- [x] I have create/modified unit tests to cover this change
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to documentation
- [ ] I have run TAF: _\<please include a link to TAF run>_  Need to merge changes for https://github.com/Breeding-Insight/bi-docker-stack/pull/20 before TAF can be run

### CODE REVIEW

**Review by HMS17 — CHANGES_REQUESTED** [review]
Testing locally revealed some problems.

Brapi:

- **Single trait add errors**
- **Batch trait import errors**
- Batch germplasm import passes
- Germplasm table display passes
- Germplasm list table display passes
- Germplasm export passes 

BB:

- First single trait add passes
- **Second single trait add "passes" but doesn't display
- Batch trait import "passes" but doesn't display and neither does second single trait. Traits do not display after waiting and refreshing, but do display once bi-api is restarted**
- Batch germplasm import passes and germplasm displayed for two separate imports
- **Displaying germplasm lists errors**

**Comment by davedrp** [comment]
Developer Testing failed - 
**Steps to reproduce**
1) viewed Ontology list (42-ontologies were present)
2) Import 3-ontologies.
3) clear browser cash
4) viewed Ontology list

**Expected result**
Ontology list has 45-ontologies (The original 42 plus the 3 new ontologies).

**Actual result**
Only the original 42-ontologies were present.  The 3 new ontologies were not present.

**NOTE**
if the bi-api server was restarted, and the Ontology list was viewed, the  3 new ontologies were not present.

**Review by davedrp — CHANGES_REQUESTED** [review]
developer-test failed

**Review by davedrp — APPROVED** [review]

**Comment by davedrp** [comment]
passed my Developer testing

**Review by HMS17 — APPROVED** [review]
Tests passed!

Brapi:
* Single trait add passes
* Batch trait import passes
* Batch germplasm import passes
* Germplasm table display passes
* Germplasm list table display passes
* Germplasm export passes

BB:
* First single trait add passes
* Second single trait add passes
* Batch trait import passes
* Batch germplasm import passes and germplasm displayed for two imports
* Displaying germplasm lists errors - but error also shows up in qa-test

File: src/main/java/org/breedinginsight/daos/impl/TraitDAOImpl.java [hunk 1]
```diff
@@ -0,0 +1,686 @@
...
+                }
+            }
+        }
+
+        return createdTraits;
+    }
+
+    @Override
+    public Trait updateTraitBrAPI(Trait trait, Program program) {
+        //TODO: Need to roll back somehow if there is an error
```
[the remark this answers is not in the document]
Comment (timparsons) [reply src/main/java/org/breedinginsight/daos/impl/TraitDAOImpl.java:466]: NA, was a pre-existing TODO

File: src/main/java/org/breedinginsight/daos/impl/TraitDAOImpl.java [hunk 2]
```diff
@@ -0,0 +1,686 @@
...
+                    //.contextOfUse()
+                    //.documentationURL()
+                    //.growthStage()
+
+            brApiVariables.add(brApiVariable);
+        }
+
+
+        // POST variables to each brapi service
+        // TODO: If there is a failure after the first brapi service, roll back all before the failure.
```
[the remark this answers is not in the document]
Comment (timparsons) [reply src/main/java/org/breedinginsight/daos/impl/TraitDAOImpl.java:425]: NA, was a pre-existing TODO

File: src/main/java/org/breedinginsight/daos/impl/TraitDAOImpl.java [hunk 3]
```diff
@@ -0,0 +1,686 @@
...
+            return Optional.empty();
+        }
+
+        return Optional.of(parseTraitRecord(record, createdByUser, updatedByUser));
+    }
+
+    @Override
+    public List<Trait> createTraitsBrAPI(List<Trait> traits, User actingUser, Program program){
+
+        //TODO: Pass ontology reference
```
[the remark this answers is not in the document]
Comment (timparsons) [reply src/main/java/org/breedinginsight/daos/impl/TraitDAOImpl.java:338]: NA, was a pre-existing TODO

File: src/main/java/org/breedinginsight/daos/impl/TraitDAOImpl.java [hunk 4]
```diff
@@ -0,0 +1,686 @@
...
+    @Override
+    public List<BrAPIObservation> getObservationsForTraitsByBrAPIProgram(String brapiProgramId, List<UUID> traitIds) {
+
+        List<String> ids = traitIds.stream()
+                .map(trait -> trait.toString())
+                .collect(Collectors.toList());
+
+        List<BrAPIObservationVariable> variables = searchVariables(ids);
+
+        // TODO: make sure have all expected external references
```
[the remark this answers is not in the document]
Comment (timparsons) [reply src/main/java/org/breedinginsight/daos/impl/TraitDAOImpl.java:287]: NA, was a pre-existing TODO

File: src/main/java/org/breedinginsight/daos/impl/TraitDAOImpl.java [hunk 5]
```diff
@@ -0,0 +1,686 @@
...
+    @Override
+    public List<BrAPIObservation> getObservationsForTraits(List<UUID> traitIds) {
+
+        List<String> ids = traitIds.stream()
+                .map(trait -> trait.toString())
+                .collect(Collectors.toList());
+
+        List<BrAPIObservationVariable> variables = searchVariables(ids);
+
+        // TODO: make sure have all expected external references
```
[the remark this answers is not in the document]
Comment (timparsons) [reply src/main/java/org/breedinginsight/daos/impl/TraitDAOImpl.java:267]: NA, was a pre-existing TODO

File: src/main/java/org/breedinginsight/daos/impl/TraitDAOImpl.java [hunk 6]
```diff
@@ -0,0 +1,686 @@
...
+
+    private List<Trait> fetchTraitsFullByProgramId(UUID programId) {
+        List<Trait> saturatedTraits = new ArrayList<>();
+        // Get our db traits (equivalent to brapi variables)
+        List<Trait> programTraits = getTraitsByProgramIds(programId);
+        if (programTraits.size() == 0) {
+            return new ArrayList<>();
+        }
+
+//        Map<UUID, Trait> dbVariablesMap = dbVariables.stream().collect(Collectors.toMap(Trait::getId, p -> p));
```
[the remark this answers is not in the document]
Comment (timparsons) [reply src/main/java/org/breedinginsight/daos/impl/TraitDAOImpl.java]: Removed

File: src/main/java/org/breedinginsight/daos/impl/ProgramDAOImpl.java [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+        BrAPIProgram brApiProgram = new BrAPIProgram()
+                                                .programName(program.getName() + " (" + program.getKey() + ")")
+                                                .abbreviation(program.getAbbreviation())
+                                                .commonCropName(program.getSpecies().getCommonName())
+                                                .externalReferences(List.of(externalReference))
+                                                .objective(program.getObjective())
+                                                .documentationURL(program.getDocumentationUrl());
+
+        // POST programs to each brapi service
+        // TODO: If there is a failure after the first brapi service, roll back all before the failure.
```
[the remark this answers is not in the document]
Comment (timparsons) [reply src/main/java/org/breedinginsight/daos/impl/ProgramDAOImpl.java:315]: NA, was a pre-existing TODO

File: src/main/java/org/breedinginsight/daos/impl/ProgramDAOImpl.java [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+                throw new HttpServerException("Could not find program in BrAPI service.");
+            }
+
+            if (brApiPrograms.getBody().getResult().getData().size() == 0){
+                throw new HttpServerException("Could not find program in BrAPI service.");
+            }
+
+            BrAPIProgram brApiProgram = brApiPrograms.getBody().getResult().getData().get(0);
+
+            //TODO: Need to add archived/not archived when available in brapi
```
[the remark this answers is not in the document]
Comment (timparsons) [reply src/main/java/org/breedinginsight/daos/impl/ProgramDAOImpl.java:355]: NA, was a pre-existing TODO

File: src/main/java/org/breedinginsight/daos/impl/ProgramDAOImpl.java [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+    }
+
+    private void initializeHttpClient(BrAPIClient brapiClient) {
+        brapiClient.setHttpClient(brapiClient.getHttpClient()
+                .newBuilder()
+                .readTimeout(getRequestTimeout())
+                .build());
+    }
+
+    //TODO figure out why BrAPIServiceFilterIntegrationTest fails when requestTimeout is set in the constructor
```
[the remark this answers is not in the document]
Comment (timparsons) [reply src/main/java/org/breedinginsight/daos/impl/ProgramDAOImpl.java:396]: NA, was a pre-existing TODO

**Pull request merged**